### PR TITLE
Documentation: Lay the ground for i18n work.

### PR DESCRIPTION
**Pull request by Spredzy**
##### SUMMARY

This pull request aims to provide the necessary entrypoints to ease dealing with localisation purposes.

Following this PR, one can:

  * Generate POT files: `make gettext`
  * Generate language specific po files `LANGUAGES=ja make generate-po`
  * List language specific po files that needs translation `LANGUAGES=ja make needs-translation` (ie. that contains either fuzzy or untranslated segment)
  
 

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME

* N/A

##### ADDITIONAL INFORMATION

* N/A

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -275,6 +278,26 @@ webdocs:
...
 	(cd docs/docsite/; CPUS=$(CPUS) $(MAKE) coredocs)
 
+.PHONY: gettext
+gettext:
+	(cd docs/docsite/; CPUS=$(CPUS) $(MAKE) gettext)
+
+.PHONY: generate-po
+generate-po:
+ifeq ($(LANGUAGES),)
+	@echo 'LANGUAGES is not defined. It is mandatory. LANGUAGES should be a comma separated list of languages to suppor. (Exampe: fr,es)'
```
Comment (abadger) [inline Makefile]:
```suggestion
	@echo 'LANGUAGES is not defined. It is mandatory. LANGUAGES should be a comma separated list of languages to support. (Exampe: fr,es)'
```

File: Makefile [hunk 2]
```diff
@@ -275,6 +278,26 @@ webdocs:
...
+else
+	(cd docs/docsite/; sphinx-intl update -w 0 -d rst/locales -p _build/gettext -l $(LANGUAGES))
+endif
+
+.PHONY: list-po-needs-translation
+list-po-needs-translation:
+ifeq ($(LANGUAGES),)
+	@echo 'LANGUAGES is not defined. It is mandatory. LANGUAGES should be a comma separated list of languages to suppor. (Exampe: fr,es)'
```
Comment (abadger) [inline Makefile]:
```suggestion
	@echo 'LANGUAGES is not defined. It is mandatory. LANGUAGES should be a comma separated list of languages to support. (Exampe: fr,es)'
```

File: docs/docsite/Makefile [hunk 3]
```diff
@@ -86,6 +90,15 @@ core_structure: core_generate_rst
 	-ln -sf ../rst/core_index.rst rst/index.rst
 	-ln -sf ../sphinx_conf/core_conf.py rst/conf.py
 
+# Symlink the relevant index into place for building core docs
+gettext_structure: gettext_generate_rst
+	@echo "Creating symlinks in gettext_generate_rst"
+	-ln -sf ../rst/core_index.rst rst/index.rst
+	-ln -sf ../sphinx_conf/core_conf.py rst/conf.py
+
+gettext: gettext_structure
```
Comment (abadger) [inline docs/docsite/Makefile:106]: I haven't absorbed what needs to be done but this doesn't seem quite right. It feels like this command should somehow switch structure depending on whether it was making core or Ansible package docs.

**Comment by abadger** [comment]
Looks good.  Couple more thoughts:

* There's a comment in the main Makefile that lists the commands and what they do.  Probably should document these new commands there as well.
* The logic for the docs/docsite build are implemented in the docs/docsite/Makefile and the toplevel Makefile just called down to that.  For ease of finding this logic, i think the generate-po and list-po-needs-translation logic should be moved there as well.
* Does the is the list Makefile target named that in tower docs? If not, it doesn't read very well as English so a better name might be preferable.  Also, it sounds like it lists strings needing translations rather than files needing translations?

File: docs/docsite/Makefile [hunk 4]
```diff
@@ -64,6 +64,10 @@ coredocs: core_htmldocs
 generate_rst: collections_meta config cli keywords plugins testing
 core_generate_rst: collections_meta config cli keywords base_plugins testing
 
+# At the moment localizing the plugins and collections is not required for the ongoing
+# localisation effort. It will come at a later time.
+gettext_generate_rst: config cli keywords testing
```
Comment (abadger) [inline docs/docsite/Makefile]:
```suggestion
gettext_generate_rst: collections_meta config cli keywords testing
```
collections_meta is information about the metadata in collections so it is in scope for this.

File: docs/docsite/Makefile.sphinx [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line.
 SPHINXCONFDIR = rst
-SPHINXOPTS    = -j $(CPUS) -n -w rst_warnings -c "$(SPHINXCONFDIR)"
+SPHINXOPTS    ?= -j $(CPUS) -n -w rst_warnings -c "$(SPHINXCONFDIR)"
```
Comment (abadger) [inline docs/docsite/Makefile.sphinx]: If this change was for step 6 here: https://www.sphinx-doc.org/en/1.8/intl.html#quick-guide we probably want to make the language parameter separate from the SPHINXOPTS and then get added to it.  Something similar to:

``` Makefile
LANGOPTS ?= -l en_US
SPHINXOPTS = -j $(CPUS) -n -w rst_warnings -c "$(SPHINXCONFDIR)"  $(LANGOPTS)
```

and then on the command line:
```
make LANGOPTS='-l pt_BR -l es_MX' html
```

File: docs/docsite/Makefile (hunk 3 above)
Comment (abadger) [reply docs/docsite/Makefile:106]: Maybe like this:
``` Makefile

gettext: ansible_structure
<TAB> CPUS=$(CPUS) $(MAKE) -f Makefile.sphinx gettext

core_gettext: core_structure   
<TAB> CPUS=$(CPUS) $(MAKE) -f Makefile.sphinx gettext
```

**Comment by Spredzy** [comment]
Thanks @abadger for the review

> There's a comment in the main Makefile that lists the commands and what they do. Probably should document these new commands there as well.

Done.

> The logic for the docs/docsite build are implemented in the docs/docsite/Makefile and the toplevel Makefile just called down to that. For ease of finding this logic, i think the generate-po and list-po-needs-translation logic should be moved there as well.

Done.

> Does the is the list Makefile target named that in tower docs? If not, it doesn't read very well as English so a better name might be preferable. Also, it sounds like it lists strings needing translations rather than files needing translations?

Rename to `needs-translation`. If not appropriate, please let me know and I'll update. (Not tight to tower docs, I can update it there too). 

It is actually printing the list of files that have untranslated or fuzzy (modified since last translation) strings.

```
#> LANGUAGES=ja make needs-translation
(cd docs/docsite/; CPUS=8 LANGUAGES=ja make needs-translation)
make[1]: Entering directory '/home/spredzy/Projects/github.com/Spredzy/ansible/ansible/docs/docsite'
(cd docs/docsite/; sphinx-intl stat -d rst/locales -l ja | grep -E ' [1-9][0-9]* (fuzzy|untranslated)' | sort)
/bin/sh: line 0: cd: docs/docsite/: No such file or directory
rst/locales/ja/LC_MESSAGES/404.po: 1 translated, 0 fuzzy, 2 untranslated.
rst/locales/ja/LC_MESSAGES/api.po: 0 translated, 0 fuzzy, 19 untranslated.
rst/locales/ja/LC_MESSAGES/cli.po: 0 translated, 0 fuzzy, 224 untranslated.
rst/locales/ja/LC_MESSAGES/collections.po: 0 translated, 0 fuzzy, 2 untranslated.
rst/locales/ja/LC_MESSAGES/community.po: 0 translated, 0 fuzzy, 682 untranslated.
rst/locales/ja/LC_MESSAGES/dev_guide.po: 0 translated, 0 fuzzy, 3554 untranslated.
rst/locales/ja/LC_MESSAGES/galaxy.po: 0 translated, 0 fuzzy, 225 untranslated.
rst/locales/ja/LC_MESSAGES/index.po: 0 translated, 0 fuzzy, 15 untranslated.
rst/locales/ja/LC_MESSAGES/installation_guide.po: 0 translated, 0 fuzzy, 158 untranslated.
rst/locales/ja/LC_MESSAGES/inventory.po: 0 translated, 0 fuzzy, 10 untranslated.
rst/locales/ja/LC_MESSAGES/network.po: 0 translated, 0 fuzzy, 1220 untranslated.
rst/locales/ja/LC_MESSAGES/plugins.po: 0 translated, 0 fuzzy, 232 untranslated.
rst/locales/ja/LC_MESSAGES/porting_guides.po: 0 translated, 0 fuzzy, 56 untranslated.
rst/locales/ja/LC_MESSAGES/reference_appendices.po: 0 translated, 0 fuzzy, 2153 untranslated.
rst/locales/ja/LC_MESSAGES/roadmap.po: 0 translated, 0 fuzzy, 49 untranslated.
rst/locales/ja/LC_MESSAGES/scenario_guides.po: 0 translated, 0 fuzzy, 1928 untranslated.
rst/locales/ja/LC_MESSAGES/user_guide.po: 0 translated, 0 fuzzy, 3344 untranslated.
make[1]: Leaving directory '/home/spredzy/Projects/github.com/Spredzy/ansible/ansible/docs/docsite'
```

File: docs/docsite/Makefile [hunk 6]
```diff
@@ -97,16 +101,19 @@ core_structure: core_generate_rst
 gettext_structure: gettext_generate_rst
 	@echo "Creating symlinks in gettext_generate_rst"
 	-ln -sf ../rst/core_index.rst rst/index.rst
-	-ln -sf ../sphinx_conf/core_conf.py rst/conf.py
+	-ln -sf ../sphinx_conf/all_conf.py rst/conf.py
 
 gettext: gettext_structure
 	CPUS=$(CPUS) $(MAKE) -f Makefile.sphinx gettext
+	# if msgcat is installed handle all indexes, otherwise use the index from gettext_structure.
+	-msgcat "$(POTDIR)/core_index.pot" "$(POTDIR)/ansible_index.pot" "$(POTDIR)/2.10_index.pot" > "$(POTDIR)/tmp_index.pot" && mv "$(POTDIR)/tmp_index.pot" "$(POTDIR)/index.pot"
```
Comment (Spredzy) [inline docs/docsite/Makefile:109]: Should we make this 2.10 the result of a dynamic command so we don't have to keep thinking about it in the future ?

Comment (abadger) [reply docs/docsite/Makefile:109]: Spredzy and I talked about this on slack.  2.10 is a special case.  We have it for backwards compatibility only. (2.10_conf.py builds nearly all of the documentation as one website.  core_conf.py builds a subset for the /ansible-core/ site.  ansible_conf.py builds a subset for /ansible/3/ and beyond).  We're not going to have a 2.11_conf.py or a 3_conf.py; those versions are taken care of by the core_conf.py and ansible_conf.py.  So we shouldn't need to get the VERSION dynamically.

**Pull request merged**